File: leet-code/206. Reverse Linked List/ReverseLinkedList206.java
public class ReverseLinkedList206 {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        cur = prev;
        return cur;
    }
}

File: leet-code/206. Reverse Linked List/ReverseLinkedList206.kt
package com.example

class ReverseLinkedList206 {
    fun reverseList(head: ListNode?): ListNode? {
        return reverse(head, null)
    }

    private fun reverse(cur: ListNode?, rev: ListNode?): ListNode? {
        if (cur == null) {
            return rev
        }

        var current = cur
        val prev = current
        current = cur.next
        prev.next = rev
        return reverse(current, prev)
    }
}